From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 17:54:04 +0900
Subject: [PATCH] Add PluginSendMessageEvent API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index dcbda5b35ab4d44e6b1d6c34bb8763a7dd08e629..5947314af6ebb448254e10e140eb2ecbe7e6ef86 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -125,6 +125,7 @@ import org.bukkit.plugin.messaging.StandardMessenger;
 import org.bukkit.scoreboard.Scoreboard;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
 
 @DelegateDeserialization(CraftOfflinePlayer.class)
 public class CraftPlayer extends CraftHumanEntity implements Player {
@@ -217,8 +218,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     @Override
     public void sendRawMessage(String message) {
         if (getHandle().playerConnection == null) return;
-
-        for (IChatBaseComponent component : CraftChatMessage.fromString(message)) {
+        // Paper start - implement PluginSendStringMessageEvent
+        PluginSendMessageEvent event = new PluginSendMessageEvent(this, message);
+        getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) return;
+        // Paper end
+        for (IChatBaseComponent component : CraftChatMessage.fromString(event.getMessage())) {
             getHandle().playerConnection.sendPacket(new PacketPlayOutChat(component));
         }
     }
