From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 17:54:04 +0900
Subject: [PATCH] Add PluginSendMessageEvent API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 0321b54f3defc983f1ebe47bf2888cbba39d102c..7dcba3ee35f08eaddc6a20801e8abd8965d1d4fd 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -128,6 +128,7 @@ import org.bukkit.plugin.messaging.StandardMessenger;
 import org.bukkit.scoreboard.Scoreboard;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
 
 @DelegateDeserialization(CraftOfflinePlayer.class)
 public class CraftPlayer extends CraftHumanEntity implements Player {
@@ -221,7 +222,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public void sendRawMessage(String message) {
         if (getHandle().playerConnection == null) return;
 
-        for (IChatBaseComponent component : CraftChatMessage.fromString(message)) {
+        // Paper start - implement PluginSendStringMessageEvent
+        PluginSendMessageEvent event = new PluginSendMessageEvent(this, message);
+        getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) return;
+        // Paper end
+        for (IChatBaseComponent component : CraftChatMessage.fromString(event.getMessage())) {
             getHandle().playerConnection.sendPacket(new PacketPlayOutChat(component, ChatMessageType.CHAT, SystemUtils.b));
         }
     }
