From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 17:56:53 +0900
Subject: [PATCH] Add PluginSendActionBarEvent API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 5947314af6ebb448254e10e140eb2ecbe7e6ef86..57c24d70f830ff60db8760a4bebab24f0225d358 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -125,6 +125,7 @@ import org.bukkit.plugin.messaging.StandardMessenger;
 import org.bukkit.scoreboard.Scoreboard;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import xyz.acrylicstyle.paper.event.server.PluginSendActionBarEvent;
 import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
 
 @DelegateDeserialization(CraftOfflinePlayer.class)
@@ -246,7 +247,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     @Override
     public void sendActionBar(String message) {
         if (getHandle().playerConnection == null || message == null || message.isEmpty()) return;
-        getHandle().playerConnection.sendPacket(new PacketPlayOutChat(new net.minecraft.server.ChatComponentText(message), net.minecraft.server.ChatMessageType.GAME_INFO));
+        // Paper start - implement PluginSendActionBarEvent
+        PluginSendActionBarEvent event = new PluginSendActionBarEvent(this, message);
+        getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) return;
+        getHandle().playerConnection.sendPacket(new PacketPlayOutChat(new net.minecraft.server.ChatComponentText(event.getMessage()), net.minecraft.server.ChatMessageType.GAME_INFO));
+        // Paper end
     }
 
     @Override
