From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 20:35:14 +0900
Subject: [PATCH] add CraftEventFactory methods


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index ce8d7877adcc4635ebf2f5743c8b02aa95691102..4ee3f2c35988d98b2c2d7f7a79ddead591b0ca3c 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -214,6 +214,8 @@ import org.bukkit.inventory.meta.BookMeta;
 import org.bukkit.potion.PotionEffect;
 
 import org.bukkit.event.entity.SpawnerSpawnEvent; // Spigot
+import xyz.acrylicstyle.paper.event.server.PluginSendActionBarEvent;
+import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
 
 public class CraftEventFactory {
     public static final DamageSource MELTING = CraftDamageSource.copyOf(DamageSource.BURN);
@@ -1714,4 +1716,18 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
         return event;
     }
+
+    // Paper start
+    public static boolean callPluginSendMessageEvent(Player player, String message) {
+        PluginSendMessageEvent event = new PluginSendMessageEvent(player, message);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+
+    public static boolean callPluginSendActionBarEvent(Player player, String message) {
+        PluginSendActionBarEvent event = new PluginSendActionBarEvent(player, message);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+    // Paper end
 }
