From ac6b0d9395282c3aa0e24111eb0805a7fb144888 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 30 Apr 2020 04:18:15 +0900
Subject: [PATCH] Add BlockLootGenerateEvent


diff --git a/src/main/java/net/minecraft/server/TileEntityLootable.java b/src/main/java/net/minecraft/server/TileEntityLootable.java
index d4cbce324..51682764d 100644
--- a/src/main/java/net/minecraft/server/TileEntityLootable.java
+++ b/src/main/java/net/minecraft/server/TileEntityLootable.java
@@ -1,5 +1,15 @@
 package net.minecraft.server;
 
+// Paper start
+import org.bukkit.Bukkit;
+import org.bukkit.block.Block;
+import org.bukkit.craftbukkit.block.CraftBarrel;
+import org.bukkit.craftbukkit.block.CraftChest;
+import org.bukkit.craftbukkit.block.CraftShulkerBox;
+import xyz.acrylicstyle.paper.event.block.BlockLootGenerateEvent;
+import java.util.stream.Collectors;
+import java.util.List;
+// Paper end
 import java.util.Random;
 import javax.annotation.Nullable;
 
@@ -60,6 +70,28 @@ public abstract class TileEntityLootable extends TileEntityContainer {
             }
 
             loottable.fillInventory(this, loottableinfo_builder.build(LootContextParameterSets.CHEST));
+
+            // Paper start - Add LootGenerateEvent (SPIGOT-2304)
+            final List<org.bukkit.inventory.ItemStack> items = getItems().stream().map(ItemStack::asBukkitCopy).collect(Collectors.toList());
+            Block block;
+            if (this instanceof TileEntityChest) {
+                block = new CraftChest(this.getBlock().createCraftBlockData().getMaterial(), (TileEntityChest) this).getBlock();
+            } else if (this instanceof TileEntityShulkerBox) {
+                block = new CraftShulkerBox(this.getBlock().createCraftBlockData().getMaterial(), (TileEntityShulkerBox) this).getBlock();
+            } else if (this instanceof TileEntityBarrel) {
+                block = new CraftBarrel(this.getBlock().createCraftBlockData().getMaterial(), (TileEntityBarrel) this).getBlock();
+            } else {
+                MinecraftServer.LOGGER.warn("Unknown TileEntityLootable class: " + this.getClass().getCanonicalName());
+                return;
+            }
+            BlockLootGenerateEvent event;
+            if (Bukkit.getServer().isPrimaryThread()) {
+                event = new BlockLootGenerateEvent(entityhuman == null ? null : entityhuman.getBukkitEntity(), block, items);
+            } else {
+                event = new BlockLootGenerateEvent(true, entityhuman == null ? null : entityhuman.getBukkitEntity(), block, items);
+            }
+            Bukkit.getPluginManager().callEvent(event);
+            // Paper end
         }
 
     }
@@ -120,6 +152,7 @@ public abstract class TileEntityLootable extends TileEntityContainer {
         this.f().clear();
     }
 
+    protected NonNullList<ItemStack> getItems() { return f(); } // Paper - OBFHELPER
     protected abstract NonNullList<ItemStack> f();
 
     protected abstract void a(NonNullList<ItemStack> nonnulllist);
-- 
2.21.0.windows.1

