From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 17:56:53 +0900
Subject: [PATCH] Add PluginSendActionBarEvent API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index e83ed45b70a97bce6cf9dd213263c93ab8eba4a8..09794bb95daa8962cc2614822c98745a57b46e8d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -130,6 +130,7 @@ import org.bukkit.plugin.messaging.StandardMessenger;
 import org.bukkit.scoreboard.Scoreboard;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import xyz.acrylicstyle.paper.event.server.PluginSendActionBarEvent;
 import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
 
 @DelegateDeserialization(CraftOfflinePlayer.class)
@@ -249,7 +250,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
 
     // Paper start
-    @Override
+    //@Override
     public void sendActionBar(BaseComponent[] message) {
         if (getHandle().playerConnection == null) return;
         getHandle().playerConnection.sendPacket(new PacketPlayOutTitle(PacketPlayOutTitle.EnumTitleAction.ACTIONBAR, message, -1, -1, -1));
@@ -258,7 +259,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     @Override
     public void sendActionBar(String message) {
         if (getHandle().playerConnection == null || message == null || message.isEmpty()) return;
-        getHandle().playerConnection.sendPacket(new PacketPlayOutTitle(PacketPlayOutTitle.EnumTitleAction.ACTIONBAR, CraftChatMessage.fromStringOrNull(message)));
+        // Paper start - implement PluginSendActionBarEvent
+        PluginSendActionBarEvent event = new PluginSendActionBarEvent(this, message);
+        getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) return;
+        getHandle().playerConnection.sendPacket(new PacketPlayOutTitle(PacketPlayOutTitle.EnumTitleAction.ACTIONBAR, CraftChatMessage.fromStringOrNull(event.getMessage())));
+        // Paper end
     }
 
     @Override
