From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 21:37:37 +0900
Subject: [PATCH] Add Server reload related events


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b89f99a66fe2ab9ad4c956c38c9e4b1d79716c9c..0c2fa88f3562b78bec6c0bb257b71525a5f80bd5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -119,6 +119,7 @@ import org.bukkit.craftbukkit.command.BukkitCommandWrapper;
 import org.bukkit.craftbukkit.command.CraftCommandMap;
 import org.bukkit.craftbukkit.command.VanillaCommandWrapper;
 import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.generator.CraftChunkData;
 import org.bukkit.craftbukkit.help.SimpleHelpMap;
 import org.bukkit.craftbukkit.inventory.CraftBlastingRecipe;
@@ -199,6 +200,8 @@ import org.yaml.snakeyaml.constructor.SafeConstructor;
 import org.yaml.snakeyaml.error.MarkedYAMLException;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import xyz.acrylicstyle.paper.event.server.ServerPreReloadEvent;
+import xyz.acrylicstyle.paper.event.server.ServerReloadedEvent;
 
 import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
@@ -778,6 +781,7 @@ public final class CraftServer implements Server {
 
     @Override
     public void reload() {
+        if (CraftEventFactory.callServerPreReloadEvent()) return; // Paper - cancel reload when the evnet was cancelled
         org.spigotmc.WatchdogThread.hasStarted = false; // Paper - Disable watchdog early timeout on reload
         reloadCount++;
         configuration = YamlConfiguration.loadConfiguration(getConfigFile());
@@ -892,6 +896,7 @@ public final class CraftServer implements Server {
         enablePlugins(PluginLoadOrder.POSTWORLD);
         getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.RELOAD));
         org.spigotmc.WatchdogThread.hasStarted = true; // Paper - Disable watchdog early timeout on reload
+        getPluginManager().callEvent(new ServerReloadedEvent()); // Paper - Add simple event to avoid bugs
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 4ee3f2c35988d98b2c2d7f7a79ddead591b0ca3c..5a44c581dee1bcc1a8d354e20f22b0d803b0955d 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -214,8 +214,10 @@ import org.bukkit.inventory.meta.BookMeta;
 import org.bukkit.potion.PotionEffect;
 
 import org.bukkit.event.entity.SpawnerSpawnEvent; // Spigot
+import xyz.acrylicstyle.paper.event.command.ReloadCommandEvent;
 import xyz.acrylicstyle.paper.event.server.PluginSendActionBarEvent;
 import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
+import xyz.acrylicstyle.paper.event.server.ServerPreReloadEvent;
 
 public class CraftEventFactory {
     public static final DamageSource MELTING = CraftDamageSource.copyOf(DamageSource.BURN);
@@ -1718,14 +1720,8 @@ public class CraftEventFactory {
     }
 
     // Paper start
-    public static boolean callPluginSendMessageEvent(Player player, String message) {
-        PluginSendMessageEvent event = new PluginSendMessageEvent(player, message);
-        Bukkit.getPluginManager().callEvent(event);
-        return event.isCancelled();
-    }
-
-    public static boolean callPluginSendActionBarEvent(Player player, String message) {
-        PluginSendActionBarEvent event = new PluginSendActionBarEvent(player, message);
+    public static boolean callServerPreReloadEvent() {
+        ServerPreReloadEvent event = new ServerPreReloadEvent();
         Bukkit.getPluginManager().callEvent(event);
         return event.isCancelled();
     }
