From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 20:35:14 +0900
Subject: [PATCH] add CraftEventFactory methods


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index d4b77d70bfc91339346689ad58f585856633007b..30018d4ff775db1b9a574162cea6696242253e61 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -216,6 +216,8 @@ import org.bukkit.potion.PotionEffect;
 import org.bukkit.projectiles.ProjectileSource;
 
 import org.bukkit.event.entity.SpawnerSpawnEvent; // Spigot
+import xyz.acrylicstyle.paper.event.server.PluginSendActionBarEvent;
+import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
 
 public class CraftEventFactory {
     public static final DamageSource MELTING = CraftDamageSource.copyOf(DamageSource.BURN);
@@ -1733,4 +1735,18 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
         return event;
     }
+
+    // Paper start
+    public static boolean callPluginSendMessageEvent(Player player, String message) {
+        PluginSendMessageEvent event = new PluginSendMessageEvent(player, message);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+
+    public static boolean callPluginSendActionBarEvent(Player player, String message) {
+        PluginSendActionBarEvent event = new PluginSendActionBarEvent(player, message);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+    // Paper end
 }
