From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 20:35:14 +0900
Subject: [PATCH] add CraftEventFactory methods


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 521e204260e96c3010fdfb90e52f7170ad9a9fc4..6531937377600b464ae5ab1c797d1cfe60c1d731 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -219,6 +219,8 @@ import org.bukkit.potion.PotionEffect;
 import org.bukkit.projectiles.ProjectileSource;
 
 import org.bukkit.event.entity.SpawnerSpawnEvent; // Spigot
+import xyz.acrylicstyle.paper.event.server.PluginSendActionBarEvent;
+import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
 
 public class CraftEventFactory {
     public static final DamageSource MELTING = CraftDamageSource.copyOf(DamageSource.BURN);
@@ -1741,4 +1743,18 @@ public class CraftEventFactory {
         StriderTemperatureChangeEvent event = new StriderTemperatureChangeEvent((Strider) strider.getBukkitEntity(), shivering);
         Bukkit.getPluginManager().callEvent(event);
     }
+
+    // Paper start
+    public static boolean callPluginSendMessageEvent(Player player, String message) {
+        PluginSendMessageEvent event = new PluginSendMessageEvent(player, message);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+
+    public static boolean callPluginSendActionBarEvent(Player player, String message) {
+        PluginSendActionBarEvent event = new PluginSendActionBarEvent(player, message);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+    // Paper end
 }
