From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Sun, 3 May 2020 10:51:40 +0900
Subject: [PATCH] disable async /fill for now


diff --git a/src/main/java/net/minecraft/server/CommandFill.java b/src/main/java/net/minecraft/server/CommandFill.java
index 8b4cc86ed87d5254ae6fa35225a1c777855d08ee..20bdebdc284c1234bace1de092ea202370556351 100644
--- a/src/main/java/net/minecraft/server/CommandFill.java
+++ b/src/main/java/net/minecraft/server/CommandFill.java
@@ -51,8 +51,8 @@ public class CommandFill {
         //if (i > 32768) {
         //    throw CommandFill.a.create(32768, i);
         //} else {
-        new Thread(() -> {
-            org.spigotmc.AsyncCatcher.enabled = false;
+        //new Thread(() -> { // disable async fill for now
+            //org.spigotmc.AsyncCatcher.enabled = false;
             List<BlockPosition> list = Lists.newArrayList();
             WorldServer worldserver = commandlistenerwrapper.getWorld();
             int j = 0;
@@ -91,8 +91,8 @@ public class CommandFill {
             } else {
                 commandlistenerwrapper.sendMessage(new ChatMessage("commands.fill.success", new Object[]{j}), true);
             }
-            org.spigotmc.AsyncCatcher.enabled = true;
-        }).start();
+            //org.spigotmc.AsyncCatcher.enabled = true;
+        //}).start();
         //}
         return 0;
         // Paper end
