From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 20:35:14 +0900
Subject: [PATCH] add CraftEventFactory methods


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 32202eb92207995b621d151481aa44bd3817e808..1ae9f31eea7065241e39baca12e2d44abf7736b1 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -225,6 +225,8 @@ import org.bukkit.potion.PotionEffect;
 import org.bukkit.projectiles.ProjectileSource;
 
 import org.bukkit.event.entity.SpawnerSpawnEvent; // Spigot
+import xyz.acrylicstyle.paper.event.server.PluginSendActionBarEvent;
+import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
 
 public class CraftEventFactory {
     public static final DamageSource MELTING = CraftDamageSource.copyOf(DamageSource.BURN);
@@ -1750,6 +1752,20 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
     }
 
+    // Paper start
+    public static boolean callPluginSendMessageEvent(Player player, String message) {
+        PluginSendMessageEvent event = new PluginSendMessageEvent(player, message);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+
+    public static boolean callPluginSendActionBarEvent(Player player, String message) {
+        PluginSendActionBarEvent event = new PluginSendActionBarEvent(player, message);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+    // Paper end
+
     public static boolean handleEntitySpellCastEvent(EntityIllagerWizard caster, EntityIllagerWizard.Spell spell) {
         EntitySpellCastEvent event = new EntitySpellCastEvent((Spellcaster) caster.getBukkitEntity(), CraftSpellcaster.toBukkitSpell(spell));
         Bukkit.getPluginManager().callEvent(event);
