From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 22:44:01 +0900
Subject: [PATCH] Fix MC-181313


diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index 9c97edf9c9e9a8cdf029264f6b563090142c686b..4b0c4da5408e5596fdd2516617b7f977d83fc246 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -347,7 +347,8 @@ public abstract class EntityArrow extends Entity implements IProjectile {
         }
 
         if (this.isCritical()) {
-            i += this.random.nextInt(i / 2 + 2);
+            int bound = i / 2 + 2;
+            i += this.random.nextInt(bound < 0 ? Integer.MAX_VALUE : bound); // Paper - Fixes MC-181313
         }
 
         Entity entity1 = this.getShooter();
