From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 22:07:49 +0900
Subject: [PATCH] Expose EntityHuman#jump to the HumanEntity API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index a2815073bc04dd831fa22da47162031d360683dd..1b26317e7a51f2cd47f01497fa1574ec7097f2b3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -646,4 +646,11 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         getHandle().openSign(teSign);
     }
     // Paper end
+
+    // Paper start - Expose HumanEntity#jump()
+    @Override
+    public void jump() {
+        getHandle().jump();
+    }
+    // Paper end
 }
