From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 21:37:37 +0900
Subject: [PATCH] Add Server reload related events


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 3bbc2d78b38f3e6353e80c7d99c16bef74a4d6c1..892d6984740d2411ba1012429ed90cfd150dc65d 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -144,6 +144,7 @@ import org.bukkit.craftbukkit.command.BukkitCommandWrapper;
 import org.bukkit.craftbukkit.command.CraftCommandMap;
 import org.bukkit.craftbukkit.command.VanillaCommandWrapper;
 import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.generator.CraftChunkData;
 import org.bukkit.craftbukkit.help.SimpleHelpMap;
 import org.bukkit.craftbukkit.inventory.CraftBlastingRecipe;
@@ -224,6 +225,7 @@ import org.yaml.snakeyaml.constructor.SafeConstructor;
 import org.yaml.snakeyaml.error.MarkedYAMLException;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import xyz.acrylicstyle.paper.event.server.ServerReloadedEvent;
 
 import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
@@ -810,6 +812,7 @@ public final class CraftServer implements Server {
 
     @Override
     public void reload() {
+        if (CraftEventFactory.callServerPreReloadEvent()) return; // Paper - cancel reload when the evnet was cancelled
         org.spigotmc.WatchdogThread.hasStarted = false; // Paper - Disable watchdog early timeout on reload
         reloadCount++;
         configuration = YamlConfiguration.loadConfiguration(getConfigFile());
@@ -930,6 +933,7 @@ public final class CraftServer implements Server {
         enablePlugins(PluginLoadOrder.POSTWORLD);
         getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.RELOAD));
         org.spigotmc.WatchdogThread.hasStarted = true; // Paper - Disable watchdog early timeout on reload
+        getPluginManager().callEvent(new ServerReloadedEvent()); // Paper - Add simple event to avoid bugs
     }
 
     // Paper start
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 6531937377600b464ae5ab1c797d1cfe60c1d731..bd3d55399ea35bc1bb2912ddb9c62e5e48a4459d 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -221,6 +221,7 @@ import org.bukkit.projectiles.ProjectileSource;
 import org.bukkit.event.entity.SpawnerSpawnEvent; // Spigot
 import xyz.acrylicstyle.paper.event.server.PluginSendActionBarEvent;
 import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
+import xyz.acrylicstyle.paper.event.server.ServerPreReloadEvent;
 
 public class CraftEventFactory {
     public static final DamageSource MELTING = CraftDamageSource.copyOf(DamageSource.BURN);
@@ -1745,14 +1746,8 @@ public class CraftEventFactory {
     }
 
     // Paper start
-    public static boolean callPluginSendMessageEvent(Player player, String message) {
-        PluginSendMessageEvent event = new PluginSendMessageEvent(player, message);
-        Bukkit.getPluginManager().callEvent(event);
-        return event.isCancelled();
-    }
-
-    public static boolean callPluginSendActionBarEvent(Player player, String message) {
-        PluginSendActionBarEvent event = new PluginSendActionBarEvent(player, message);
+    public static boolean callServerPreReloadEvent() {
+        ServerPreReloadEvent event = new ServerPreReloadEvent();
         Bukkit.getPluginManager().callEvent(event);
         return event.isCancelled();
     }
