From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 10 Jun 2020 13:20:03 +0900
Subject: [PATCH] Figure out why InventoryClickEvent isn't calling - drop when
 resolved


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 76c9c89eb66a6264dcdfa77058db03ac31c6f795..9feff5508621789621012fb7c9da713a566638be 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -2235,6 +2235,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
             } else {
                 // CraftBukkit start - Call InventoryClickEvent
                 if (packetplayinwindowclick.c() < -1 && packetplayinwindowclick.c() != -999) {
+                    System.out.println("Invalid slot, skipping firing InventoryClickEvent"); // Paper - TODO: figure out why InventoryClickEvent isn't calling
                     return;
                 }
 
@@ -2499,7 +2500,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
                         // correctly support custom recipes
                         player.updateInventory(player.activeContainer);
                     }
-                }
+                } else System.out.println("Quick craft, ignoring"); // Paper - TODO: figure out why InventoryClickEvent isn't calling
                 // CraftBukkit end
                 if (ItemStack.matches(packetplayinwindowclick.f(), itemstack)) {
                     this.player.playerConnection.sendPacket(new PacketPlayOutTransaction(packetplayinwindowclick.b(), packetplayinwindowclick.e(), true));
