From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 15 May 2020 23:45:56 +0900
Subject: [PATCH] Fix sending quit message twice


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 7035d586ab76dc7960db39100a3addc02fd6b8f4..45c254b777d73d44c9573b46a405d9c249e510da 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1600,7 +1600,9 @@ public class PlayerConnection implements PacketListenerPlayIn {
                 }
             });
             // Paper end
-            this.minecraftServer.getPlayerList().sendMessage(CraftChatMessage.fromString(quitMessage.getQuitMessage()));
+            for (IChatBaseComponent iChatBaseComponent : chatMessage) {
+                this.minecraftServer.sendMessage(iChatBaseComponent);
+            }
         }
         // CraftBukkit end
         if (this.isExemptPlayer()) {
