From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Sun, 24 May 2020 01:25:41 +0900
Subject: [PATCH] Add TileEntity API


diff --git a/src/main/java/net/minecraft/server/TileEntity.java b/src/main/java/net/minecraft/server/TileEntity.java
index a8e64dfdab1e73894144a65c10c15d22f9198d3d..0a25e6d5898a9bb1e2ac2850583006e155bed8b1 100644
--- a/src/main/java/net/minecraft/server/TileEntity.java
+++ b/src/main/java/net/minecraft/server/TileEntity.java
@@ -168,6 +168,7 @@ public abstract class TileEntity implements KeyedObject { // Paper
         return null;
     }
 
+    public NBTTagCompound save() { return b(); } // Paper - OBFHELPER
     public NBTTagCompound b() {
         return this.d(new NBTTagCompound());
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index efaf86eeb9f98eba07d4277c76edbba77a59215e..67e4f27511e92ff1ff903720431b040ea9819d05 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -43,43 +43,7 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 import javax.imageio.ImageIO;
 //import jline.console.ConsoleReader; // Paper
-import net.minecraft.server.Advancement;
-import net.minecraft.server.ArgumentEntity;
-import net.minecraft.server.Block;
-import net.minecraft.server.BlockPosition;
-import net.minecraft.server.BossBattleCustom;
-import net.minecraft.server.CommandDispatcher;
-import net.minecraft.server.CommandListenerWrapper;
-import net.minecraft.server.DedicatedPlayerList;
-import net.minecraft.server.DedicatedServer;
-import net.minecraft.server.DedicatedServerProperties;
-import net.minecraft.server.DedicatedServerSettings;
-import net.minecraft.server.DimensionManager;
-import net.minecraft.server.Enchantments;
-import net.minecraft.server.EntityPlayer;
-import net.minecraft.server.EnumDifficulty;
-import net.minecraft.server.EnumGamemode;
-import net.minecraft.server.IRecipe;
-import net.minecraft.server.IRegistry;
-import net.minecraft.server.Item;
-import net.minecraft.server.ItemWorldMap;
-import net.minecraft.server.Items;
-import net.minecraft.server.JsonListEntry;
-import net.minecraft.server.LootTableRegistry;
-import net.minecraft.server.MapIcon;
-import net.minecraft.server.MinecraftKey;
-import net.minecraft.server.MobEffects;
-import net.minecraft.server.PlayerList;
-import net.minecraft.server.ServerCommand;
-import net.minecraft.server.TagsServer;
-import net.minecraft.server.TicketType;
-import net.minecraft.server.Vec3D;
-import net.minecraft.server.WorldData;
-import net.minecraft.server.WorldMap;
-import net.minecraft.server.WorldNBTStorage;
-import net.minecraft.server.WorldServer;
-import net.minecraft.server.WorldSettings;
-import net.minecraft.server.WorldType;
+import net.minecraft.server.*;
 import org.apache.commons.lang.Validate;
 import org.apache.commons.lang3.StringUtils;
 import org.bukkit.BanList;
@@ -2274,5 +2238,10 @@ public final class CraftServer implements Server {
     public com.destroystokyo.paper.entity.ai.MobGoals getMobGoals() {
         return mobGoals;
     }
+
+    @Override
+    public xyz.acrylicstyle.paper.block.TileEntity createTileEntity(xyz.acrylicstyle.paper.nbt.NBTTagCompound nbtTagCompound) {
+        return new xyz.acrylicstyle.paper.block.CraftTileEntity(TileEntity.create((NBTTagCompound) xyz.acrylicstyle.paper.nbt.CraftNBT.asNMSCopy(nbtTagCompound)));
+    }
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 7230ddbf6eb765390543bdb3ff8c08d383bb2666..c26738d3cef00ddbc9860c6683c43621ef47896b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -22,78 +22,7 @@ import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
-import net.minecraft.server.ArraySetSorted;
-import net.minecraft.server.AxisAlignedBB;
-import net.minecraft.server.BiomeBase;
-import net.minecraft.server.BiomeDecoratorGroups;
-import net.minecraft.server.BlockChorusFlower;
-import net.minecraft.server.BlockDiodeAbstract;
-import net.minecraft.server.BlockPosition;
-import net.minecraft.server.Blocks;
-import net.minecraft.server.ChunkCoordIntPair;
-import net.minecraft.server.ChunkMapDistance;
-import net.minecraft.server.ChunkStatus;
-import net.minecraft.server.EntityAreaEffectCloud;
-import net.minecraft.server.EntityArmorStand;
-import net.minecraft.server.EntityArrow;
-import net.minecraft.server.EntityBoat;
-import net.minecraft.server.EntityEgg;
-import net.minecraft.server.EntityEnderSignal;
-import net.minecraft.server.EntityEvokerFangs;
-import net.minecraft.server.EntityExperienceOrb;
-import net.minecraft.server.EntityFallingBlock;
-import net.minecraft.server.EntityFireball;
-import net.minecraft.server.EntityFireworks;
-import net.minecraft.server.EntityHanging;
-import net.minecraft.server.EntityHuman;
-import net.minecraft.server.EntityInsentient;
-import net.minecraft.server.EntityItem;
-import net.minecraft.server.EntityItemFrame;
-import net.minecraft.server.EntityLeash;
-import net.minecraft.server.EntityLightning;
-import net.minecraft.server.EntityMinecartChest;
-import net.minecraft.server.EntityMinecartCommandBlock;
-import net.minecraft.server.EntityMinecartFurnace;
-import net.minecraft.server.EntityMinecartHopper;
-import net.minecraft.server.EntityMinecartMobSpawner;
-import net.minecraft.server.EntityMinecartRideable;
-import net.minecraft.server.EntityMinecartTNT;
-import net.minecraft.server.EntityPainting;
-import net.minecraft.server.EntityPotion;
-import net.minecraft.server.EntitySnowball;
-import net.minecraft.server.EntityTNTPrimed;
-import net.minecraft.server.EntityTippedArrow;
-import net.minecraft.server.EntityTypes;
-import net.minecraft.server.EntityZombie;
-import net.minecraft.server.EnumDifficulty;
-import net.minecraft.server.EnumDirection;
-import net.minecraft.server.EnumMobSpawn;
-import net.minecraft.server.ExceptionWorldConflict;
-import net.minecraft.server.Explosion;
-import net.minecraft.server.GameRules;
-import net.minecraft.server.GroupDataEntity;
-import net.minecraft.server.IBlockData;
-import net.minecraft.server.IChunkAccess;
-import net.minecraft.server.MCUtil;
-import net.minecraft.server.MinecraftKey;
-import net.minecraft.server.MinecraftServer;
-import net.minecraft.server.MovingObjectPosition;
-import net.minecraft.server.PacketPlayOutCustomSoundEffect;
-import net.minecraft.server.PacketPlayOutUpdateTime;
-import net.minecraft.server.PacketPlayOutWorldEvent;
-import net.minecraft.server.PersistentRaid;
-import net.minecraft.server.PlayerChunk;
-import net.minecraft.server.ProtoChunkExtension;
-import net.minecraft.server.RayTrace;
-import net.minecraft.server.SoundCategory;
-import net.minecraft.server.Ticket;
-import net.minecraft.server.TicketType;
-import net.minecraft.server.Unit;
-import net.minecraft.server.Vec3D;
-import net.minecraft.server.WorldGenerator;
-import net.minecraft.server.WorldProvider;
-import net.minecraft.server.WorldProviderTheEnd;
-import net.minecraft.server.WorldServer;
+import net.minecraft.server.*; // Paper
 import org.apache.commons.lang.Validate;
 import org.bukkit.BlockChangeDelegate;
 import org.bukkit.Bukkit;
@@ -2519,6 +2448,44 @@ public class CraftWorld implements World {
     }
     // Paper end - per player view distance
 
+    // Paper start - TileEntity API
+    @Override
+    public xyz.acrylicstyle.paper.block.TileEntity getTileEntity(Location location) {
+        return getTileEntity(location.getBlockX(), location.getBlockY(), location.getBlockZ());
+    }
+
+    @Override
+    public xyz.acrylicstyle.paper.block.TileEntity getTileEntity(int x, int y, int z) {
+        net.minecraft.server.TileEntity te = getHandle().getTileEntity(new BlockPosition(x, y, z));
+        if (te == null) return null;
+        if (te instanceof TileEntityContainer) {
+            return new xyz.acrylicstyle.paper.block.CraftTileEntityContainer((TileEntityContainer) te);
+        } else {
+            return new xyz.acrylicstyle.paper.block.CraftTileEntity(te);
+        }
+    }
+
+    @Override
+    public void removeTileEntity(Location location) {
+        removeTileEntity(location.getBlockX(), location.getBlockY(), location.getBlockZ());
+    }
+
+    @Override
+    public void removeTileEntity(int x, int y, int z) {
+        getHandle().removeTileEntity(new BlockPosition(x, y, z));
+    }
+
+    @Override
+    public void setTileEntity(Location location, xyz.acrylicstyle.paper.block.TileEntity tileEntity) {
+        setTileEntity(location.getBlockX(), location.getBlockY(), location.getBlockZ(), tileEntity);
+    }
+
+    @Override
+    public void setTileEntity(int x, int y, int z, xyz.acrylicstyle.paper.block.TileEntity tileEntity) {
+        getHandle().setTileEntity(new BlockPosition(x, y, z), ((xyz.acrylicstyle.paper.block.CraftTileEntity) tileEntity).getHandle());
+    }
+    // Paper end - TileEntity API
+
     // Spigot start
     private final Spigot spigot = new Spigot()
     {
diff --git a/src/main/java/xyz/acrylicstyle/paper/block/CraftTileEntity.java b/src/main/java/xyz/acrylicstyle/paper/block/CraftTileEntity.java
new file mode 100644
index 0000000000000000000000000000000000000000..487505a8903c5a46de90d528c773bc76a47cd79d
--- /dev/null
+++ b/src/main/java/xyz/acrylicstyle/paper/block/CraftTileEntity.java
@@ -0,0 +1,71 @@
+package xyz.acrylicstyle.paper.block;
+
+import net.minecraft.server.BlockPosition;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.block.Block;
+import org.bukkit.inventory.InventoryHolder;
+import xyz.acrylicstyle.paper.nbt.CraftNBT;
+import xyz.acrylicstyle.paper.nbt.NBTTagCompound;
+
+public class CraftTileEntity implements TileEntity {
+    net.minecraft.server.TileEntity handle;
+
+    public CraftTileEntity(net.minecraft.server.TileEntity handle) {
+        this.handle = handle;
+    }
+
+    public net.minecraft.server.TileEntity getHandle() {
+        return handle;
+    }
+
+    public void setHandle(net.minecraft.server.TileEntity handle) {
+        this.handle = handle;
+    }
+
+    @Override
+    public World getWorld() {
+        return handle.getWorld() == null ? null : handle.getWorld().getWorld();
+    }
+
+    @Override
+    public Location getLocation() {
+        BlockPosition p = handle.getPosition();
+        return new Location(handle.getWorld() == null ? null : handle.getWorld().getWorld(), p.getX(), p.getY(), p.getZ());
+    }
+
+    @Override
+    public void load(NBTTagCompound nbtTagCompound) {
+        handle.load(CraftNBT.asNMSCompound(nbtTagCompound));
+    }
+
+    @Override
+    public NBTTagCompound save(NBTTagCompound nbtTagCompound) {
+        return CraftNBT.asBukkitCompound(handle.save(CraftNBT.asNMSCompound(nbtTagCompound)));
+    }
+
+    @Override
+    public void update() {
+        handle.update();
+    }
+
+    @Override
+    public Block getBlock() {
+        return getLocation().getBlock();
+    }
+
+    @Override
+    public NBTTagCompound save() {
+        return CraftNBT.asBukkitCompound(handle.save());
+    }
+
+    @Override
+    public boolean isRemoved() {
+        return handle.isRemoved();
+    }
+
+    @Override
+    public InventoryHolder getOwner() {
+        return handle.getOwner();
+    }
+}
diff --git a/src/main/java/xyz/acrylicstyle/paper/block/CraftTileEntityContainer.java b/src/main/java/xyz/acrylicstyle/paper/block/CraftTileEntityContainer.java
new file mode 100644
index 0000000000000000000000000000000000000000..72cc9439d7cc03152d665764e7d9ea8fbd6b67da
--- /dev/null
+++ b/src/main/java/xyz/acrylicstyle/paper/block/CraftTileEntityContainer.java
@@ -0,0 +1,43 @@
+package xyz.acrylicstyle.paper.block;
+
+import net.minecraft.server.ChatComponentText;
+import org.bukkit.Location;
+
+public class CraftTileEntityContainer extends CraftTileEntity implements TileEntityContainer {
+    net.minecraft.server.TileEntityContainer handle;
+
+    public CraftTileEntityContainer(net.minecraft.server.TileEntityContainer handle) {
+        super(handle);
+        this.handle = handle;
+    }
+
+    public void setHandle(net.minecraft.server.TileEntityContainer handle) {
+        super.setHandle(handle);
+        this.handle = handle;
+    }
+
+    @Override
+    public net.minecraft.server.TileEntityContainer getHandle() {
+        return handle;
+    }
+
+    @Override
+    public String getDisplayName() {
+        return handle.getDisplayName().getText();
+    }
+
+    @Override
+    public String getCustomName() {
+        return handle.getCustomName() == null ? null : handle.getCustomName().getText();
+    }
+
+    @Override
+    public void setCustomName(String name) {
+        handle.setCustomName(new ChatComponentText(name));
+    }
+
+    @Override
+    public Location getLocation() {
+        return handle.getLocation();
+    }
+}
diff --git a/src/main/java/xyz/acrylicstyle/paper/nbt/CraftNBT.java b/src/main/java/xyz/acrylicstyle/paper/nbt/CraftNBT.java
index fbb18fb6208869b768c3f427e6fa4c467047c66e..0273687878857d417bec75a3c9aeb013437254f0 100644
--- a/src/main/java/xyz/acrylicstyle/paper/nbt/CraftNBT.java
+++ b/src/main/java/xyz/acrylicstyle/paper/nbt/CraftNBT.java
@@ -32,6 +32,10 @@ public final class CraftNBT {
         return l;
     }
 
+    public static net.minecraft.server.NBTTagCompound asNMSCompound(NBTTagCompound nbt) {
+        return (net.minecraft.server.NBTTagCompound) asNMSCopy(nbt);
+    }
+
     public static NBTBase asNMSCopy(xyz.acrylicstyle.paper.nbt.NBTBase nbt) {
         if (nbt instanceof NBTTagString) {
             return net.minecraft.server.NBTTagString.a(nbt.asString());
@@ -63,6 +67,10 @@ public final class CraftNBT {
         } else throw new IllegalArgumentException("Unknown type: " + nbt.getClass().getCanonicalName());
     }
 
+    public static NBTTagCompound asBukkitCompound(net.minecraft.server.NBTTagCompound nbt) {
+        return (NBTTagCompound) asBukkitCopy(nbt);
+    }
+
     public static xyz.acrylicstyle.paper.nbt.NBTBase asBukkitCopy(NBTBase nbt) {
         if (nbt instanceof net.minecraft.server.NBTTagString) {
             return NBTTagString.create(nbt.asString());
