From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 30 Apr 2020 04:18:15 +0900
Subject: [PATCH] Add BlockLootGenerateEvent


diff --git a/src/main/java/net/minecraft/server/TileEntityLootable.java b/src/main/java/net/minecraft/server/TileEntityLootable.java
index 55b1e5da2b5128e479dbdf14eb1da692af1880fe..abab1e5a1495082adaaed8af06d7ad98c8cc1095 100644
--- a/src/main/java/net/minecraft/server/TileEntityLootable.java
+++ b/src/main/java/net/minecraft/server/TileEntityLootable.java
@@ -1,5 +1,17 @@
 package net.minecraft.server;
 
+// Paper start
+import org.bukkit.Bukkit;
+import org.bukkit.block.Block;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.block.CraftBarrel;
+import org.bukkit.craftbukkit.block.CraftChest;
+import org.bukkit.craftbukkit.block.CraftShulkerBox;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import xyz.acrylicstyle.paper.event.block.BlockLootGenerateEvent;
+import java.util.stream.Collectors;
+import java.util.List;
+// Paper end
 import java.util.Random;
 import javax.annotation.Nullable;
 
@@ -65,6 +77,31 @@ public abstract class TileEntityLootable extends TileEntityContainer {
             }
 
             loottable.fillInventory(this, loottableinfo_builder.build(LootContextParameterSets.CHEST));
+
+            // Paper start - Add LootGenerateEvent (SPIGOT-2304)
+            ((CraftServer) Bukkit.getServer()).getServer().postToMainThread(() -> {
+                final List<org.bukkit.inventory.ItemStack> items = getItems().stream().map(ItemStack::asBukkitCopy).collect(Collectors.toList());
+                Block block;
+                if (TileEntityLootable.this instanceof TileEntityChest) {
+                    block = new CraftChest(this.getBlock().createCraftBlockData().getMaterial(), (TileEntityChest) this).getBlock();
+                } else if (TileEntityLootable.this instanceof TileEntityShulkerBox) {
+                    block = new CraftShulkerBox(this.getBlock().createCraftBlockData().getMaterial(), (TileEntityShulkerBox) this).getBlock();
+                } else if (TileEntityLootable.this instanceof TileEntityBarrel) {
+                    block = new CraftBarrel(this.getBlock().createCraftBlockData().getMaterial(), (TileEntityBarrel) this).getBlock();
+                } else {
+                    MinecraftServer.LOGGER.warn("Unknown TileEntityLootable class: " + this.getClass().getCanonicalName());
+                    return;
+                }
+                BlockLootGenerateEvent event;
+                if (Bukkit.getServer().isPrimaryThread()) {
+                    event = new BlockLootGenerateEvent(entityhuman == null ? null : entityhuman.getBukkitEntity(), block, items);
+                } else {
+                    event = new BlockLootGenerateEvent(true, entityhuman == null ? null : entityhuman.getBukkitEntity(), block, items);
+                }
+                Bukkit.getPluginManager().callEvent(event);
+                setItems(new NonNullList<>(event.getGeneratedLoot().stream().map(CraftItemStack::asNMSCopy).collect(Collectors.toList()), null));
+            });
+            // Paper end
         }
 
     }
@@ -132,8 +169,10 @@ public abstract class TileEntityLootable extends TileEntityContainer {
         this.f().clear();
     }
 
+    protected NonNullList<ItemStack> getItems() { return f(); } // Paper - OBFHELPER
     protected abstract NonNullList<ItemStack> f();
 
+    protected void setItems(NonNullList<ItemStack> items) { a(items); } // Paper - OBFHELPER
     protected abstract void a(NonNullList<ItemStack> nonnulllist);
 
     @Override
