From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 15 May 2020 23:45:56 +0900
Subject: [PATCH] Fix sending quit message twice


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 1d3972f42a1268addf4abe4f81d6cfa25be080ae..db085d3b144a0cb44db070872815cb03a4422693 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1538,7 +1538,9 @@ public class PlayerConnection implements PacketListenerPlayIn {
                 }
             });
             // Paper end
-            this.minecraftServer.getPlayerList().sendMessage(CraftChatMessage.fromString(quitMessage.getQuitMessage()));
+            for (IChatBaseComponent iChatBaseComponent : chatMessage) {
+                this.minecraftServer.sendMessage(iChatBaseComponent);
+            }
         }
         // CraftBukkit end
         if (this.isExemptPlayer()) {
