From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 15 May 2020 23:45:56 +0900
Subject: [PATCH] Fix sending quit message twice


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 9bc98c05b3e5112e75a9d312d78368a2331cd11e..1473fcf431d7052e7d88a1a74b483fb73e17d1b7 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1604,7 +1604,9 @@ public class PlayerConnection implements PacketListenerPlayIn {
                 }
             });
             // Paper end
-            this.minecraftServer.getPlayerList().sendMessage(CraftChatMessage.fromString(quitMessage.getQuitMessage()));
+            for (IChatBaseComponent iChatBaseComponent : chatMessage) {
+                this.minecraftServer.sendMessage(iChatBaseComponent);
+            }
         }
         // CraftBukkit end
         if (this.isExemptPlayer()) {
