From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 15 May 2020 23:45:56 +0900
Subject: [PATCH] Fix sending quit message twice


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 5db425d19dcd9c8e2def6b6a74273e739c635423..abd88123466735a02cb1eb43619ca8a9a99e4cb8 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1599,7 +1599,9 @@ public class PlayerConnection implements PacketListenerPlayIn {
                 }
             });
             // Paper end
-            this.minecraftServer.getPlayerList().sendMessage(CraftChatMessage.fromString(quitMessage.getQuitMessage()));
+            for (IChatBaseComponent iChatBaseComponent : chatMessage) {
+                this.minecraftServer.sendMessage(iChatBaseComponent);
+            }
         }
         // CraftBukkit end
         if (this.isExemptPlayer()) {
