From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 19 Aug 2020 00:18:44 +0900
Subject: [PATCH] Add and Implement PlayerDeathEvent Recipients


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 73ea9518316ff465e86fe95143875c9de5aa2fa8..be59f7982cca87ca1c63eb6af2616d9b90c84b0f 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -714,7 +714,10 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
                     this.server.getPlayerList().b(this, ichatbasecomponent);
                 }
             } else {
-                this.server.getPlayerList().sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, SystemUtils.b);
+                // Grid start - implement PlayerDeathEvent recipients
+                event.getRecipients().forEach(player -> ((CraftPlayer) player).getHandle().sendMessage(ichatbasecomponent, SystemUtils.b));
+                // this.server.getPlayerList().sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, SystemUtils.b);
+                // Grid end - implement PlayerDeathEvent recipients
             }
         } else {
             this.playerConnection.sendPacket(new PacketPlayOutCombatEvent(this.getCombatTracker(), PacketPlayOutCombatEvent.EnumCombatEventType.ENTITY_DIED));
