From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 19 Aug 2020 00:19:12 +0900
Subject: [PATCH] Add BlockPlaceInteractEvent


diff --git a/src/main/java/net/minecraft/server/ItemBlock.java b/src/main/java/net/minecraft/server/ItemBlock.java
index bbb1de7d1adce6e986467db234b27518df4dd87f..1a6b0a8f049491b7369cadb45d59115f8812b6f8 100644
--- a/src/main/java/net/minecraft/server/ItemBlock.java
+++ b/src/main/java/net/minecraft/server/ItemBlock.java
@@ -61,6 +61,16 @@ public class ItemBlock extends Item {
                         block.postPlace(world, blockposition, iblockdata1, entityhuman, itemstack);
                         // CraftBukkit start
                         if (blockstate != null) {
+                            // Grid start - implement BlockPlaceInteractEvent
+                            /* // disable for now
+                            xyz.acrylicstyle.paper.event.block.BlockPlaceInteractEvent placeInteractEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockPlaceInteractEvent(entityhuman, blockstate, blockactioncontext1.getHand());
+                            if (placeInteractEvent.isCancelBlockPlace()) {
+                                blockstate.update(true, false);
+                            }
+                            if (placeInteractEvent.isCancelItemUsage()) {
+                                return EnumInteractionResult.FAIL;
+                            }*/
+                            // Grid end - implement BlockPlaceInteractEvent
                             org.bukkit.event.block.BlockPlaceEvent placeEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockPlaceEvent((WorldServer) world, entityhuman, blockactioncontext1.getHand(), blockstate, blockposition.getX(), blockposition.getY(), blockposition.getZ());
                             if (placeEvent != null && (placeEvent.isCancelled() || !placeEvent.canBuild())) {
                                 blockstate.update(true, false);
diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index 3adb29f004d2fee36f3ee9b21ee5417e84b64837..6a9d468f8939428fe0ffe2ff8a0494fd6894b75e 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -37,6 +37,7 @@ import org.bukkit.entity.Player;
 import org.bukkit.event.block.BlockFertilizeEvent;
 import org.bukkit.event.player.PlayerItemDamageEvent;
 import org.bukkit.event.world.StructureGrowEvent;
+//import xyz.acrylicstyle.paper.event.block.BlockPlaceInteractEvent; // Grid
 // CraftBukkit end
 
 public final class ItemStack {
@@ -273,13 +274,41 @@ public final class ItemStack {
             if (entityhuman != null && enuminteractionresult.a()) {
                 org.bukkit.event.block.BlockPlaceEvent placeEvent = null;
                 List<BlockState> blocks = new java.util.ArrayList<>(world.capturedBlockStates.values());
+                xyz.acrylicstyle.paper.event.block.BlockPlaceInteractEvent placeInteractEvent =
+                    blocks.size() >= 1 ? org.bukkit.craftbukkit.event.CraftEventFactory.callBlockPlaceInteractEvent(entityhuman, blocks.get(0), enumhand) : null; // Grid
                 world.capturedBlockStates.clear();
                 if (blocks.size() > 1) {
                     placeEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockMultiPlaceEvent(world, entityhuman, enumhand, blocks, blockposition.getX(), blockposition.getY(), blockposition.getZ());
                 } else if (blocks.size() == 1) {
                     placeEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockPlaceEvent(world, entityhuman, enumhand, blocks.get(0), blockposition.getX(), blockposition.getY(), blockposition.getZ());
                 }
+                // Grid start - implement BlockPlaceInteractEvent
+                boolean cancelPlaceInteract = false;
+                if (placeInteractEvent != null) {
+                    if (placeInteractEvent.isCancelBlockPlace()) { // copied from below
+                        cancelPlaceInteract = true;
+                        // PAIL: Remove this when MC-99075 fixed
+                        placeInteractEvent.getPlayer().updateInventory();
+                        world.capturedTileEntities.clear(); // Paper - clear out tile entities as chests and such will pop loot
+                        // revert back all captured blocks
+                        for (BlockState blockstate : blocks) {
+                            blockstate.update(true, false);
+                        }
 
+                        // Brute force all possible updates
+                        BlockPosition placedPos = ((CraftBlock) placeInteractEvent.getBlock()).getPosition();
+                        for (EnumDirection dir : EnumDirection.values()) {
+                            ((EntityPlayer) entityhuman).playerConnection.sendPacket(new PacketPlayOutBlockChange(world, placedPos.shift(dir)));
+                        }
+                    }
+                    if (placeInteractEvent.isCancelItemUsage()) {
+                        cancelPlaceInteract = true;
+                        enuminteractionresult = EnumInteractionResult.FAIL;
+                        placeInteractEvent.getPlayer().updateInventory();
+                    }
+                }
+                // Grid end - implement BlockPlaceInteractEvent
+                if (!cancelPlaceInteract) { // Grid
                 if (placeEvent != null && (placeEvent.isCancelled() || !placeEvent.canBuild())) {
                     enuminteractionresult = EnumInteractionResult.FAIL; // cancel placement
                     // PAIL: Remove this when MC-99075 fixed
@@ -362,6 +391,7 @@ public final class ItemStack {
 
                     entityhuman.b(StatisticList.ITEM_USED.b(item));
                 }
+                } // Grid
             }
             world.capturedTileEntities.clear();
             world.capturedBlockStates.clear();
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index c3fc9608e7c398b3a869ffe9853f67905d8c2c4b..6ca79e7bce48467de1c92d4498667aa6fc41837c 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -225,6 +225,7 @@ import org.bukkit.potion.PotionEffect;
 import org.bukkit.projectiles.ProjectileSource;
 
 import org.bukkit.event.entity.SpawnerSpawnEvent; // Spigot
+import xyz.acrylicstyle.paper.event.block.BlockPlaceInteractEvent;
 import xyz.acrylicstyle.paper.event.server.PluginSendActionBarEvent;
 import xyz.acrylicstyle.paper.event.server.PluginSendMessageEvent;
 import xyz.acrylicstyle.paper.event.server.ServerPreReloadEvent;
@@ -1780,6 +1781,22 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
         return event.isCancelled();
     }
+
+    public static BlockPlaceInteractEvent callBlockPlaceInteractEvent(EntityHuman entityHuman, BlockState block, EnumHand hand) {
+        Player player = ((EntityPlayer) entityHuman).getBukkitEntity();
+        org.bukkit.inventory.ItemStack item;
+        EquipmentSlot equipmentSlot;
+        if (hand == EnumHand.MAIN_HAND) {
+            item = player.getInventory().getItemInMainHand();
+            equipmentSlot = EquipmentSlot.HAND;
+        } else {
+            item = player.getInventory().getItemInOffHand();
+            equipmentSlot = EquipmentSlot.OFF_HAND;
+        }
+        BlockPlaceInteractEvent event = new BlockPlaceInteractEvent(player, block.getBlock(), item, equipmentSlot);
+        Bukkit.getPluginManager().callEvent(event);
+        return event;
+    }
     // Paper end
 
     public static boolean handleEntitySpellCastEvent(EntityIllagerWizard caster, EntityIllagerWizard.Spell spell) {
