From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 19 Aug 2020 00:18:44 +0900
Subject: [PATCH] Add and Implement PlayerDeathEvent Recipients


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 8b1e866b3e0362d3a391586bbf14702dc25b8c00..b05972d55cb9eb7d7ccecff0a57743462e6d3aea 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -714,7 +714,10 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
                     this.server.getPlayerList().b(this, ichatbasecomponent);
                 }
             } else {
-                this.server.getPlayerList().sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, SystemUtils.b);
+                // Grid start - implement PlayerDeathEvent recipients
+                event.getRecipients().forEach(player -> ((CraftPlayer) player).getHandle().sendMessage(ichatbasecomponent, SystemUtils.b));
+                // this.server.getPlayerList().sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, SystemUtils.b);
+                // Grid end - implement PlayerDeathEvent recipients
             }
         } else {
             this.playerConnection.sendPacket(new PacketPlayOutCombatEvent(this.getCombatTracker(), PacketPlayOutCombatEvent.EnumCombatEventType.ENTITY_DIED));
