From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 28 Aug 2020 13:08:15 +0900
Subject: [PATCH] Fix NPE on Explosion


diff --git a/src/main/java/net/minecraft/server/Explosion.java b/src/main/java/net/minecraft/server/Explosion.java
index b7cfa6fb188d36ecbad0389bb51e5f1016b9eabd..094783afcfed06018c9fcda873e5ab4fa6e02dd3 100644
--- a/src/main/java/net/minecraft/server/Explosion.java
+++ b/src/main/java/net/minecraft/server/Explosion.java
@@ -138,7 +138,7 @@ public class Explosion {
                                 // Paper start - prevent headless pistons from forming
                                 if (!com.destroystokyo.paper.PaperConfig.allowHeadlessPistons && iblockdata.getBlock() == Blocks.MOVING_PISTON) {
                                     TileEntityPiston extension = (TileEntityPiston)this.world.getTileEntity(blockposition);
-                                    if (extension.isHead()) {
+                                    if (extension != null && extension.isHead()) { // Grid
                                        EnumDirection direction = iblockdata.get(BlockPistonExtension.FACING);
                                        set.add(blockposition.shift(direction.opposite()));
                                     }
