From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Thu, 2 Jan 2020 12:25:16 -0600
Subject: [PATCH] Add effect to block break naturally


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 596f1d68b69a2ba3976398d0045379f047541f40..a6fe268b28591e7e964741094e7f639d6c23b52b 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -460,6 +460,18 @@ public interface Block extends Metadatable {
      */
     boolean breakNaturally();
 
+    // Paper start
+    /**
+     * Breaks the block and spawns items as if a player had digged it with a
+     * specific tool
+     *
+     * @param tool The tool or item in hand used for digging
+     * @param triggerEffect Play the block break particle effect and sound
+     * @return true if the block was destroyed
+     */
+    boolean breakNaturally(@NotNull ItemStack tool, boolean triggerEffect);
+    // Paper end
+
     /**
      * Breaks the block and spawns items as if a player had digged it with a
      * specific tool
