From 9fab6bbebdb7f361df4307081e4c986a76732fdf Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 22:07:50 +0900
Subject: [PATCH] Expose EntityHuman#jump to the HumanEntity API


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 79569a32..c44d0b73 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -486,4 +486,11 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
      */
     void openSign(@NotNull org.bukkit.block.Sign sign);
     // Paper end
+
+    // Paper start - Expose EntityHuman#jump
+    /**
+     * Force the player to jump.
+     */
+    void jump();
+    // Paper end
 }
-- 
2.21.0.windows.1

