From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Sun, 3 May 2020 10:51:30 +0900
Subject: [PATCH] refined /reload tab completion


diff --git a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
index 31afb14e96d8c7593e6411bcd5884d547ee02be7..0d6552714f89c762427941d95fe4d004c49fd488 100644
--- a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
@@ -63,9 +63,14 @@ public class ReloadCommand extends BukkitCommand {
         return true;
     }
 
+    private static final List<String> arguments = com.google.common.collect.Lists.newArrayList("permissions", "commands"); // Paper
+
     @NotNull
     @Override
     public List<String> tabComplete(@NotNull CommandSender sender, @NotNull String alias, @NotNull String[] args) throws IllegalArgumentException {
-        return com.google.common.collect.Lists.newArrayList("permissions", "commands"); // Paper
+        // Paper start - allow refined tab completion
+        if (args.length == 0) return arguments;
+        return arguments.stream().filter(s -> s.toLowerCase().startsWith(args[0].toLowerCase())).collect(java.util.stream.Collectors.toList());
+        // Paper end
     }
 }
