From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 29 Apr 2020 22:07:50 +0900
Subject: [PATCH] Expose EntityHuman#jump to the HumanEntity API


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index e6cdd4bc2d934ab932651e17f144d7004835bdc3..c7da9577eb91e60b22cef577c2363fc4fa34ef9b 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -474,4 +474,11 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
      */
     void openSign(@NotNull org.bukkit.block.Sign sign);
     // Paper end
+
+    // Paper start - Expose EntityHuman#jump
+    /**
+     * Force the player to jump.
+     */
+    void jump();
+    // Paper end
 }
